File: webserviceutil/src/main/java/com/example/webserviceutil/OkHttp/CountingFileRequestBody.java
package com.example.webserviceutil.OkHttp;

import java.io.File;
import java.io.IOException;

import okhttp3.MediaType;
import okhttp3.RequestBody;
import okhttp3.internal.Util;
import okio.BufferedSink;
import okio.Okio;
import okio.Source;

public class CountingFileRequestBody extends RequestBody {

    private static final int SEGMENT_SIZE = 1024*256; // okio.Segment.SIZE

    private final File file;
    private final ProgressListener progressListener;
    private final String contentType;

    public CountingFileRequestBody(File file, String contentType, ProgressListener progressListener) {
        this.file = file;
        this.contentType = contentType;
        this.progressListener = progressListener;
    }

    @Override
    public long contentLength() {
        return file.length();
    }

    @Override
    public MediaType contentType() {
        return MediaType.parse(contentType);
    }

    @Override
    public void writeTo(BufferedSink sink) throws IOException {
        Source source = null;
        try {
            source = Okio.source(file);
            long read;
            long segmentReaded = 0L;

            while ((read = source.read(sink.buffer(), SEGMENT_SIZE)) != -1) {
                segmentReaded += read;
                if(segmentReaded >= SEGMENT_SIZE) {
                    sink.flush();
                    progressListener.onUpdate(segmentReaded, file.getName());
                    segmentReaded = 0L;
                }
            }
            if(segmentReaded != 0) {
                sink.flush();
                progressListener.onUpdate(segmentReaded, file.getName());
                segmentReaded = 0L;
            }


        } finally {
            Util.closeQuietly(source);
        }
    }

    interface ProgressListener {
        void onUpdate(long bytesRead, String fileName);
    }

}

File: networkrequest/src/main/java/com/anima/networkrequest/data/okhttp/CountingFileRequestBody.kt
package com.anima.networkrequest.data.okhttp

import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody
import okhttp3.internal.closeQuietly
import okio.BufferedSink
import okio.Source
import okio.source
import java.io.File

/**
 * Created by jianjianhong on 20-3-11
 */
class CountingFileRequestBody( var file: File, var contentType: String, var progressListener: ProgressListener) : RequestBody(){
    private val SEGMENT_SIZE = 1024 * 256 // okio.Segment.SIZE

    override fun contentLength(): Long {
        return file.length()
    }

    override fun contentType(): MediaType? {
        return contentType.toMediaTypeOrNull()
    }

    override fun writeTo(sink: BufferedSink) {
        var source: Source? = null
        try {
            source = file.source()
            var segmentReaded = 0L

            var read = source!!.read(sink.buffer(), SEGMENT_SIZE.toLong())
            while (read != -1L) {
                segmentReaded += read
                if (segmentReaded >= SEGMENT_SIZE) {
                    sink.flush()
                    progressListener.onUpdate(segmentReaded, file.name)
                    segmentReaded = 0L
                }
                read = source!!.read(sink.buffer(), SEGMENT_SIZE.toLong())
            }
            if (segmentReaded != 0L) {
                sink.flush()
                progressListener.onUpdate(segmentReaded, file.name)
                segmentReaded = 0L
            }


        } finally {
            source!!.closeQuietly()
        }
    }

    interface ProgressListener {
        fun onUpdate(bytesRead: Long, fileName: String)
    }
}